File: day7.jl
import Base.Iterators:partition

function load_data(filename)

    open(filename) do f
        bags = Dict{AbstractString,Vector{AbstractString}}()
        for line in readlines(f)
            line = strip(replace(line, [',','.'] => ""))
            line = strip(replace(line, "bags" => ""))
            line = strip(replace(line, "bag" => ""))
            line = strip(replace(line, "no other" => ""))
            color, contain = split(strip(line),"contain")
            loads = String[]
            for x in partition(split(contain), 3)
                n , c1, c2 = x
                for i in 1:parse(Int, n)
                    push!(loads,string(c1," ",c2))
                end
            end
            bags[strip(color)] = loads
        end
        return bags
    end

end

function hold_shiny_gold( color, bags )

    if "shiny gold" in bags[color]
       return true
    elseif length(bags[color]) > 0
       return any([hold_shiny_gold( c, bags ) for c in unique(bags[color])])
    else
	   return false
    end

end

function contained_bags( color, bags, n)

    if length(bags[color]) == 0
        return n
    else
        s = length(bags[color])
        for c in bags[color]
           s += contained_bags(c, bags, n)
        end
        return s
    end


end

bags = load_data("input7.txt")
println(count([hold_shiny_gold(c, bags) for c in keys(bags)]))
println(contained_bags( "shiny gold", bags, 0))

File: day7.py


def load_data(filename):

    with open(filename) as f:
        bags = dict()
        for line in f:
            line = line.replace(',', '').replace('.', '').replace('bags','')
            line = line.replace("bag", '').replace("no other", "")
            line = line.strip()
            color, contain = line.split("contain")
            loads = []
            contain = contain.split()
            i = 0
            while i < len(contain): 
                 n = int(contain[i])
                 i += 1
                 c1 = contain[i]
                 i += 1
                 c2 = contain[i]
                 i += 1
                 loads += [f"{c1} {c2}" for j in range(n)]
            color = color.strip()
            bags[color] = loads
        return bags

def hold_shiny_gold( color, bags ):

    if "shiny gold" in bags[color]:
        return True
    elif len(bags[color]) > 0:
        return any(hold_shiny_gold( c, bags ) for c in set(bags[color]))
    else:
        return False

def contained_bags( color, bags, n):

    if len(bags[color]) == 0:
        return n
    else:
        s = len(bags[color])
        for c in bags[color]:
           s += contained_bags(c, bags, n)
        return s

bags = load_data("input7.txt")
print(sum([hold_shiny_gold(c, bags) for c in bags.keys()]))
print(contained_bags( "shiny gold", bags, 0))
